File: conway.py
import pygame
import random
pygame.init()

screen_length = 150 #python can't handle 300 (max)
fps = 60
run = True
cell_size = 1
chance = 5
pause = False
grid_size = screen_length / cell_size
itr = range(int(grid_size))
g = grid_size - 1

screen = pygame.display.set_mode((screen_length, screen_length), pygame.SCALED)
clock = pygame.time.Clock()
cells = [[] for a in itr]

for i in itr:
	for j in itr:
		d = bool(random.randint(0, chance))
		cells[i].append([d, d])
		
while run:
	for event in pygame.event.get():
		if event.type == pygame.QUIT:
			run = False
		if event.type == pygame.KEYDOWN:
			if event.key == pygame.K_SPACE:
				pause = not pause
			if event.key == pygame.K_RETURN:
				for i in cells:
					for j in i:
						j[0] = j[1] = bool(random.randint(0, chance))
								
	if not pause:	
		screen.fill((0, 0, 0))
		
		for x, i in enumerate(cells):
			for y, j in enumerate(i):
				if not j[0]:
					screen.set_at((x, y), (255, 255, 255))
				
				if (x > 0 and x < g and y > 0 and y < g):
					alive = (8 - cells[x - 1][y - 1][0] -
	 							cells[x][y - 1][0] -
	 							cells[x + 1][y - 1][0] -
	 							cells[x - 1][y][0] -
	 							cells[x + 1][y][0] -
	 							cells[x - 1][y + 1][0] -
	 							cells[x][y + 1][0] -
	 							cells[x + 1][y + 1][0])
					if not j[0]:
						if not (4 > alive > 1):
							j[1] = True
					else:
						if alive == 3:
							j[1] = False
				else:
					if not j[0]:
						j[1] = j[0] = True
				
		for i in cells:
			for j in i:
				j[0] = j[1]
				
	clock.tick(fps)
	print(clock.get_fps())
	pygame.display.update()
pygame.quit()
						

File: conway.cpp
#include <iostream>
#include <SDL2/SDL.h>
#include <SDL2/SDL_timer.h>
#include <vector>
#define fps 60

using namespace std;

struct cell
{
	bool dead;
	bool buff_dead;
};

int main(int argc, char *argv[])
{
	float fps_ms = 1000.0 / fps; //60
	float scale = 3;

	int screen_length = 300;
	int cell_size = 1; //5
	int chance = 2;	   //20
	int i, j, alive;
	int grid_size = screen_length / cell_size;

	bool run = true;
	bool pause = 0;

	SDL_Init(SDL_INIT_EVERYTHING);
	SDL_Window *screen = SDL_CreateWindow("Conway", SDL_WINDOWPOS_CENTERED, SDL_WINDOWPOS_CENTERED, screen_length, screen_length, SDL_WINDOW_FULLSCREEN);
	Uint32 render_flags = SDL_RENDERER_ACCELERATED;
	SDL_Renderer *renderer = SDL_CreateRenderer(screen, -1, render_flags);
	//SDL_RenderSetScale(renderer, scale, scale);
	SDL_RenderSetLogicalSize(renderer, screen_length, screen_length);

	srand(time(nullptr));

	Uint32 totalFrameTicks = 0;
	Uint32 totalFrames = 0;
	Uint32 endTicks;
	Uint64 endPerf;
	Uint64 framePerf;
	Uint32 startTicks;
	Uint64 startPerf;

	vector<vector<cell>> cells(grid_size);

	for (i = 0; i < grid_size; i++)
	{
		cells[i] = vector<cell>(grid_size);
		for (j = 0; j < grid_size; j++)
		{
			cells[i][j].dead = rand() % chance;
			cells[i][j].buff_dead = cells[i][j].dead;
		}
	}

	while (run)
	{
		totalFrames++;
		startTicks = SDL_GetTicks();
		startPerf = SDL_GetPerformanceCounter();
		SDL_Event event;
		while (SDL_PollEvent(&event))
		{
			switch (event.type)
			{
			case SDL_QUIT:
				run = false;
				break;

			case SDL_KEYDOWN:
				switch (event.key.keysym.sym)
				{
				case SDLK_SPACE:
					pause = not pause;
					break;

				case SDLK_RETURN:
					for (i = 0; i < grid_size; i++)
					{
						for (j = 0; j < grid_size; j++)
						{
							cells[i][j].dead = rand() % chance;
							cells[i][j].buff_dead = cells[i][j].dead;
						}
					}
					break;

				default:
					break;
				}
				break;

			default:
				break;
			}
		}

		//draw
		SDL_SetRenderDrawColor(renderer, 0, 0, 0, 255);
		SDL_RenderClear(renderer);
		SDL_SetRenderDrawColor(renderer, 255, 255, 255, 255);

		for (i = 0; i < grid_size; i++)
		{
			for (j = 0; j < grid_size; j++)
			{
				cells[i][j].dead = cells[i][j].buff_dead;
			}
		}

		for (i = 0; i < grid_size; i++)
		{
			for (j = 0; j < grid_size; j++)
			{
				if (not cells[i][j].dead)
				{
					SDL_RenderDrawPoint(renderer, i, j);
				}
				if (not pause)
				{
					if (i > 0 and i < grid_size - 1 and j > 0 and j < grid_size - 1)
					{
						alive = 8 - cells[i - 1][j - 1].dead -
								cells[i][j - 1].dead -
								cells[i + 1][j - 1].dead -
								cells[i - 1][j].dead -
								cells[i + 1][j].dead -
								cells[i - 1][j + 1].dead -
								cells[i][j + 1].dead -
								cells[i + 1][j + 1].dead;

						if (not cells[i][j].dead)
						{
							if (not(alive > 1 and alive < 4))
							{
								cells[i][j].buff_dead = true;
							}
						}
						else
						{
							if (alive == 3)
							{
								cells[i][j].buff_dead = false;
							}
						}
					}

					else
					{
						if (not cells[i][j].dead)
						{
							cells[i][j].buff_dead = true;
						}
					}
				}
			}
		}

		SDL_RenderPresent(renderer);

		endTicks = SDL_GetTicks();
		endPerf = SDL_GetPerformanceCounter();
		framePerf = (endPerf - startPerf) / (float)SDL_GetPerformanceFrequency() * 1000.0f;
		cout << (1.0f / ((endPerf - startPerf) / (float)SDL_GetPerformanceFrequency())) << endl;
		SDL_Delay(floor(fps_ms - framePerf));
	}

	SDL_DestroyRenderer(renderer);
	SDL_DestroyWindow(screen);
	SDL_Quit();
	return 0;
}
